vcpkg/src/lib.rs

```diff
@@ -398,19 +398,196 @@ fn find_vcpkg_target(cfg: &Config, target_triplet: &TargetTriplet) -> Result<Vcp
+            files.insert(pc_file.id.to_owned(), pc_file);
+        }
+        Ok(PcFiles { files: files })
+    }
+    /// Use the .pc files as a hint to the library sort order.
+    fn fix_ordering(&self, mut libs: Vec<String>) -> Vec<String> {
+        // Overall heuristic: for each library given as input, identify which PcFile declared it.
+        // Then, looking at that PcFile, check its Requires: (deps), and if the pc file for that
+        // dep is in our set, check if its libraries are in our set of libs.  If so, move it to the
+        // end to ensure it gets linked afterwards.
+
+        // We may need to do this a few times to properly handle the case where A -> (depends on) B
+        // -> C -> D and libraries were originally sorted D, C, B, A.  Avoid recursion so we don't
+        // have to detect potential cycles.
+        let mut tries_left = 3;
+        loop {
+            let mut required_lib_order: Vec<String> = Vec::new();
+            for lib in &libs {
+                required_lib_order.push(lib.to_owned());
+                if let Some(pc_file) = self.locate_pc_file_by_lib(lib) {
+                    // Consider its requirements:
+                    for dep in &pc_file.deps {
+                        // Only consider pkgconfig dependencies we know about.
+                        let dep_pc_file = match self.files.get(dep) {
+                            None => continue,
+                            Some(pc_file) => pc_file,
+                        };
+                        // Intra-port library ordering found, pivot any already seen dep_lib to the
+                        // end of the list.
+                        for dep_lib in &dep_pc_file.libs {
+                            if let Some(removed) = remove_item(&mut required_lib_order, dep_lib) {
+                                required_lib_order.push(removed);
+                            }
+                        }
+                    }
+                }
+            }
+            // We should always end up with the same number of libraries, only their order should
+            // change.
+            assert_eq!(libs.len(), required_lib_order.len());
+            // Termination:
+            if required_lib_order == libs {
+                // Nothing changed, we're done here.
+                return libs;
+            } else {
+                // Only try a few times.
+                if tries_left <= 0 {
+                    println!("cargo:warning=vcpkg gave up trying to resolve pkg-config ordering.");
+                    return libs;
+                } else {
+                    libs = required_lib_order;
+                    tries_left = tries_left - 1;
+                    continue;
+                }
+            }
+        }
+        // unreachable
+    }
+    /// Locate which PcFile contains this library, if any.
+    fn locate_pc_file_by_lib(&self, lib: &str) -> Option<&PcFile> {
+        for (id, pc_file) in &self.files {
```

Comment:
Hmm, I worry a bit about cases where the `Libs:` line contains `-l` flags for libraries that are not provided by the `.pc` file in question ... but I don't think there's any realistic way to detect that case.

Reply:
Yes other `-l` flags are possible and will parse into the PcFile, but that should be fine without any sort of validation as the sorting would only applies to the Port local set of libs gathered via the manifest, so would otherwise be ignored.

Tests include `-lm` which I've seen a few times already.